ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -523,16 +598,75 @@ private PlanNode buildJoin(final Join root, final String prefix) {
           leaf.getSource(), prefix + "Right", root.getInfo().getRightJoinExpression());
     }
 
+    final boolean finalJoin = prefix.isEmpty();
+
+    final JoinKey joinKey = buildJoinKey(root);
+
     return new JoinNode(
         new PlanNodeId(prefix + "Join"),
         root.getInfo().getType(),
+        joinKey.rewriteWith(refRewriter::process),
+        finalJoin,
         left,
         right,
         root.getInfo().getWithinExpression()
     );
   }
 
-  private DataSourceNode buildNonJoinNode(final AliasedDataSource dataSource) {
+  private JoinKey buildJoinKey(final Join join) {
```

Comment:
To be able to determine the schema of the `JoinNode` we must first know which of the join expressions, if any, will be used as the key column in the result.   This is a bit of a chicken and egg problem, and this isn't a great solution.

I have thoughts on how this can be resolved in a follow up PR.